src/BooksDisplayer/BooksDisplayer.tsx

```diff
@@ -1,189 +1,236 @@
-                   } else {
-                      modifiedFormats[key] = book.formats[key];
-                   }
-                }
-                return { ...book, formats: modifiedFormats };
-             });
-             setRecords(modifiedData);
-             setIsLoading(false);
-          })
-          .catch(error => {
-             console.error(error);
-             setIsLoading(false);
-          });
-   }, [])
-
-   //Filtering books by categories/name
-   useEffect(() => {
-      const filteredBooks = filterByCategories(records, category);
-      setFilteredBooks(filteredBooks);
-      setExpandedBook(null);
-   }, [category, records, favorites]);
-
-   useEffect(() => {
-      const filteredBooks = filterByName(records, name);
-      setFilteredBooks(filteredBooks);
-      setExpandedBook(null);
-   }, [name, records]);
-   
-   function filterByCategories(records, category){
+  const [records, setRecords] = useState<Book[]>([]);
+  const [expandedBook, setExpandedBook] = useState<number | null>(null);
+  const [filteredBooks, setFilteredBooks] = useState<Book[]>([]);
+  const [favorites, setFavorites] = useState<Book[]>([]);
+  const [category, setCategory] = useState("All");
+  const [name, setName] = useState("");
+  const [isLoading, setIsLoading] = useState(true);
+
+  interface Book {
+    id: number;
+    title: string;
+    authors: Array<{ name: string; birth_year?: number; death_year?: number }>;
+    subjects: string[];
+    bookshelves: string[];
+    copyright: boolean;
+    download_count: number;
+    formats: { [format: string]: string };
+    languages: string[];
+    media_type: string;
+    translators: unknown[];
+  }
+  // Fetching data from API
+  useEffect(() => {
+    setIsLoading(true);
+    fetch("https://gutendex.com/books")
+      .then((response) => response.json())
+      .then((data: { results: Book[] }) => {
+        setRecords(data.results);
+        setIsLoading(false);
+      })
+      .catch((error) => {
+        setIsLoading(false);
+        throw new Error(error as string);
+      });
+  }, []);
+  //Filtering books by categories/name
+  useEffect(() => {
+    function filterByCategories() {
       if (category === "All") {
-         return records;
-     }
-
-     if (category === "Favorite"){
-      return favorites;
-     }
- 
-     const regex = new RegExp(category);
- 
-     const filteredBooks = records.filter(book => {
-         const subjects = book.subjects || [];
-         return subjects.some(subject => regex.test(subject));
-     });
- 
-     return filteredBooks;
-   }
-
-
-   function filterByName(records, name){
+        return records;
+      }
+      if (category === "Favorite") {
+        return favorites;
+      }
+      const regex = new RegExp(category);
+
+      const outputFilteredBooks = records.filter((book) => {
+        const subjects = book.subjects;
+        return subjects.some((subject: string) => regex.test(subject));
```

Comment:
nie potrzebujesz tutaj adnotacji
```suggestion
        return subjects.some((subject) => regex.test(subject));
```